Update mutator interface to accept proto objects directly

Liamsi opened the pull request:
This PR changes 
```golang
type Mutation interface {
    Mutate(value, mutation []byte ([]byte, error)
}
``` 
to
```golang
type Mutation interface {
    Mutate(value, mutation proto.Message) ([]byte, error)
}
```

This PR also removes `Replacer` which isn't used anywhere.

gdbelvin commented on core/mutator/mutator.go:
```diff
@@ -47,9 +47,9 @@ var (
-	// CheckMutation verifies that this is a valid mutation for this item and
+	// Mutate verifies that this is a valid mutation for this item and
 	// applies mutation to value.
-	Mutate(value, mutation []byte) ([]byte, error)
```
To make this more generic, should we consider making the mutation a generic protobuf?

Liamsi replied:
like `proto.Any`?

gdbelvin replied:
like `proto.Message`

Liamsi replied:
Done

Liamsi commented:
We currently `objecthash.ObjectHash(nil)` if the previous entry does not exist. I've added a little helper function `FromLeafValue` to have a central place to consistently convert what we have in `LeafValue`.
PTAL

gdbelvin commented on core/client/kt/requests.go:
```diff
@@ -75,7 +77,7 @@ func (v *Verifier) CreateUpdateEntryRequest(
 		return nil, err
 	}
-	previous := objecthash.ObjectHash(getResp.GetLeafProof().GetLeaf().GetLeafValue())
+	previous := objecthash.ObjectHash(prevEntry)
```
Please add a TODO to change this hash to plain SHA256. 
The previous entry is stored as `[]byte` and therefore we should hash it directly.

Liamsi replied:
Thanks, I've added a TODO. 
I'm not sure what the second part of the comment means: `prevEntry` is of type  `Entry` and not []byte. Should I add a comment about this as well?

gdbelvin commented on core/mutator/entry/entry_test.go:
```diff
@@ -71,18 +72,21 @@ func createEntry(commitment []byte, pkeys []string) ([]byte, error) {
-func prepareMutation(key []byte, entryData []byte, previous []byte, signers []signatures.Signer) ([]byte, error) {
+func prepareMutation(key []byte, newEntry *tpb.Entry, previous []byte, signers []signatures.Signer) (*tpb.SignedKV, error) {
+	var entryData []byte
+	if newEntry != nil {
```
It should not be valid to set `newEntry` with `nil`. Please return an error in this case.

gdbelvin commented on core/mutator/entry/entry_test.go:
```diff
@@ -183,13 +182,35 @@ func TestCheckMutation(t *testing.T) {
+		want    *tpb.Entry
+		wantErr bool
+	}{
+		{make([]byte, 0), nil, false},            // empty leaf -> nil, no error
```
Add a test case for both `nil` and `[]byte{}` and make sure they produce different values

Liamsi commented:
Thanks for your review! I hope I addressed all your comments (besides the 2nd part of https://github.com/google/keytransparency/pull/739#discussion_r132683130). PTAL

gdbelvin commented on core/mutator/entry/entry.go:
```diff
@@ -41,25 +42,41 @@ func New() *Entry {
 	// Ensure that the mutation size is within bounds.
-	if proto.Size(update) > mutator.MaxMutationSize {
-		glog.Warningf("mutation (%v bytes) is larger than the maximum accepted size (%v bytes).", proto.Size(update), mutator.MaxMutationSize)
+	if proto.Size(updateM) > mutator.MaxMutationSize {
```
go style typically discourages putting type letters in variable names.

gdbelvin commented on core/mutator/entry/entry.go:
```diff
@@ -41,25 +42,41 @@ func New() *Entry {
+		}
 		// Check if this mutation is a replay.
-		if bytes.Equal(oldValue, update.GetKeyValue().Value) {
+		if bytes.Equal(entryData, update.GetKeyValue().Value) {
```
`proto.Marshal` is not required to produce consistent byte output.  Proto's binary encoding for example, could reorder tags and values.  Could this be upgraded to a `proto.Equal`?

Liamsi commented:
Thanks for the review! I've updated to `proto.Equal` and addressed the other comments. PTAL

gdbelvin commented on core/mutator/entry/entry.go:
```diff
@@ -41,51 +42,75 @@ func New() *Entry {
+	}
+	var oldEntry *tpb.Entry
+	if oldValue != nil {
+		oldEntry, ok = oldValue.(*tpb.Entry)
```
`oldEntry` shadows the outer variable.  It's safer to declare a new variable name, and then set set `outer = inner`
